src/main/java/org/folio/services/instance/InstanceEffectiveValuesService.java

```diff
@@ -0,0 +1,74 @@
+package org.folio.services.instance;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static org.folio.services.instance.PublicationPeriodParser.parsePublicationPeriod;
+
+import java.util.List;
+import java.util.Objects;
+import org.folio.rest.jaxrs.model.Instance;
+import org.folio.rest.jaxrs.model.Publication;
+import org.folio.services.batch.BatchOperationContext;
+
+public class InstanceEffectiveValuesService {
+
+  public Instance populateEffectiveValues(Instance instance) {
+    if (instance.getPublicationPeriod() == null) {
+      populatePublicationPeriod(instance);
+    }
+
+    return instance;
+  }
+
+  public void populateEffectiveValues(Instance newInstance, Instance oldInstance) {
+    if (instancePublicationsChanged(newInstance, oldInstance)) {
+      populateEffectiveValues(newInstance);
+    }
+  }
+
+  public void populateEffectiveValues(List<Instance> allInstances,
+    BatchOperationContext<Instance> context) {
+
+    var existingInstancesMap = context.getExistingRecords().stream()
+      .collect(toMap(Instance::getId, identity()));
+
+    for (Instance newInstance : allInstances) {
+      var instanceId = newInstance.getId();
+      if (existingInstancesMap.containsKey(instanceId)) {
+        populateEffectiveValues(newInstance, existingInstancesMap.get(instanceId));
+      } else {
+        populateEffectiveValues(newInstance);
+      }
+    }
+  }
+
+  private boolean instancePublicationsChanged(Instance newInstance, Instance oldInstance) {
+    var newPublications = newInstance.getPublication();
+    var oldPublications = oldInstance.getPublication();
+
+    if (newPublications == null || oldPublications == null) {
+      return true;
+    }
+
+    if (newPublications.size() != oldPublications.size()) {
+      return true;
+    }
+
+    if (newPublications.size() == 0) {
+      return false;
+    }
```

Comment:
Seems to me you can adjust the logic to return false if newPublications is null or size == 0.

Reply:
It is already returns false if newPublications is empty. If any is null than we still have to recalculate the property to remove the old value.